File: clients/typescript/src/timeout.ts
import { assert } from "./utils/common.ts";

export enum TimeoutUnit {
    Hour = "H",
    Minute = "M",
    Second = "S",
    Millisecond = "m",
}

export class Timeout {
    private constructor(
        public readonly unit: TimeoutUnit,
        public readonly value: number,
    ) { }

    static hour(hours: number): Timeout {
        return new Timeout(TimeoutUnit.Hour, hours);
    }

    static minute(mins: number): Timeout {
        return new Timeout(TimeoutUnit.Minute, mins);
    }

    static second(secs: number): Timeout {
        return new Timeout(TimeoutUnit.Second, secs);
    }

    static millisecond(ms: number): Timeout {
        return new Timeout(TimeoutUnit.Millisecond, ms);
    }

    toString(): string {
        return `${this.value}${this.unit}`;
    }

    duration(): number {
        switch (this.unit) {
            case TimeoutUnit.Hour:
                return this.value * 60 * 60 * 1000;

            case TimeoutUnit.Minute:
                return this.value * 60 * 1000;

            case TimeoutUnit.Second:
                return this.value * 1000;

            case TimeoutUnit.Millisecond:
                return this.value;
        }
    }

    static fromString(input: string): Timeout {
        assert(input.length >= 2, SyntaxError, "timeout: invalid format");

        const numPart = input.slice(0, -1);
        const unit = input.slice(-1);

        const value = Number.parseInt(numPart, 10);

        assert(Number.isFinite(value), TypeError, "timeout: invalid number");

        switch (unit) {
            case "H":
                return Timeout.hour(value);

            case "M":
                return Timeout.minute(value);

            case "S":
                return Timeout.second(value);

            case "m":
                return Timeout.millisecond(value);

            default:
                throw new Error("timeout: unknown unit");
        }
    }

    equals(other: Timeout): boolean {
        console.log({ other });
        return this.unit === other.unit && this.value === other.value;
    }
}

File: clients/dart/lib/timeout.dart
import 'package:setu_client/utils/common.dart';

enum TimeoutUnit {
  hour('H'),
  minute('M'),
  second('S'),
  millisecond('m');

  const TimeoutUnit(this.symbol);

  final String symbol;
}

class Timeout {
  factory Timeout.fromString(String input) {
    assertExpr(
      input.length >= 2,
      error: FormatException.new,
      message: () => 'timeout: invalid format',
    );

    final numPart = input.substring(0, input.length - 1);
    final unit = input.substring(input.length - 1);

    final value = int.tryParse(numPart);

    if (value == null) {
      throw FormatException('timeout: invalid number');
    }

    return switch (unit) {
      'H' => Timeout.hour(value),
      'M' => Timeout.minute(value),
      'S' => Timeout.second(value),
      'm' => Timeout.millisecond(value),
      _ => throw FormatException('timeout: unknown unit'),
    };
  }

  const Timeout._(this.unit, this.value);

  final TimeoutUnit unit;
  final int value;

  const Timeout.hour(int hours) : this._(TimeoutUnit.hour, hours);
  const Timeout.minute(int minutes) : this._(TimeoutUnit.minute, minutes);
  const Timeout.second(int seconds) : this._(TimeoutUnit.second, seconds);
  const Timeout.millisecond(int milliseconds)
    : this._(TimeoutUnit.millisecond, milliseconds);

  @override
  String toString() => '$value${unit.symbol}';

  Duration get duration => switch (unit) {
    TimeoutUnit.hour => Duration(hours: value),
    TimeoutUnit.minute => Duration(minutes: value),
    TimeoutUnit.second => Duration(seconds: value),
    TimeoutUnit.millisecond => Duration(milliseconds: value),
  };

  @override
  bool operator ==(Object other) =>
      other is Timeout && other.unit == unit && other.value == value;

  @override
  int get hashCode => Object.hash(unit, value);
}
